src/actions.jl

```diff
@@ -1,32 +1,107 @@
+abstract type ByLinearTransformation <: Action end
 
-struct ExtensionHomomorphism{T,I<:Integer,V,Op} <: AbstractActionHomomorphism{T}
-    features::V
+abstract type InducedActionHomomorphism{T} end
+
+function induce(ac::Action, hom::InducedActionHomomorphism, g::GroupElement)
+    throw(
+        "No fallback is provided for $(typeof(ac)). You need to implement `induce(::$(typeof(ac)), ::$(typeof(hom)), ::$(typeof(g)))`.",
+    )
+end
+
+decompose(x::Any, hom::InducedActionHomomorphism) = throw(
+    "No fallback is provided for $(typeof(x)). You need to implement `decompose(::$(typeof(x)), ::$(typeof(hom)))`.",
+)
+
+coeff_type(ac::ByLinearTransformation) = throw(
+    "No fallback is provided for $(typeof(ac)). You need to implement `coeff_type(::$(typeof(ac)))`.",
+)
+
+#=
+implements:
+* features(action_hom)
+* action(hom::InducedActionHomomorphism) → Action
+* action(hom::InducedActionHomomorphism, g::GroupElement, feature) → return action of `g` on `feature`
+=#
+
+struct ExtensionHomomorphism{T,A,I,V} <: InducedActionHomomorphism{T}
+    ac::A
+    features::V # supports linear indexing
     reversef::Dict{T,I}
-    op::Op
 end
 
-function ExtensionHomomorphism{I}(features, op) where I<:Integer
+@static if VERSION < v"1.3.0"
+    # cannot add methods to an abstract type in pre Julia v1.3
+    (hom::ExtensionHomomorphism)(g::GroupElement) = induce(action(hom), hom, g)
+
+    function (hom::ExtensionHomomorphism)(orb::O) where {T,O<:AbstractOrbit{T,Nothing}}
+        elts = hom.(orb)
+        dict = Dict(e => nothing for e in elts)
+        return Orbit(elts, dict)
+    end
+
+    function (hom::ExtensionHomomorphism)(chars::AbstractArray{<:Character})
+        χ = first(chars)
+        iccG = hom.(conjugacy_classes(χ))
+        return [Character(values(χ), χ.inv_of, iccG) for χ in chars]
+    end
+else
+    (hom::InducedActionHomomorphism)(g::GroupElement) = induce(action(hom), hom, g)
+
+    function (hom::InducedActionHomomorphism)(orb::O) where {T,O<:AbstractOrbit{T,Nothing}}
+        elts = hom.(orb)
+        dict = Dict(e => nothing for e in elts)
+        return Orbit(elts, dict)
+    end
+
+    function (hom::InducedActionHomomorphism)(chars::AbstractArray{<:Character})
+        χ = first(chars)
+        iccG = hom.(conjugacy_classes(χ))
+        return [Character(values(χ), χ.inv_of, iccG) for χ in chars]
+    end
+end
+
+function ExtensionHomomorphism{I}(ac::Action, features) where {I<:Integer}
     @assert typemax(I) >= length(features) "Use wider Integer type!"
-    reversef = Dict{eltype(features), I}(f => idx for (idx, f) in pairs(features))
-    return ExtensionHomomorphism(features, reversef, op)
+    reversef = Dict{eltype(features),I}(f => idx for (idx, f) in pairs(features))
+    return ExtensionHomomorphism(ac, features, reversef)
 end
 
-ExtensionHomomorphism(features, op) = ExtensionHomomorphism{Int16}(features, op)
+ExtensionHomomorphism(ac::Action, features) = ExtensionHomomorphism{Int16}(ac, features)
```

Comment:
Why `Int16` and not `Int64` ?

Reply:
probably `UInt16` would be better; I went for smaller type than Int to save space;

Exceeding `typemax(UInt16)` here would mean e.g. that you're trying to block-diagonalize SDP constraint of size 65535×65535, which is highly unlikely ;)

Reply:
Makes sense, we could add a comment saying that :)